### PR TITLE
Single line ignore switches

### LINKED ISSUE #144: Ignore a line with a single comment
Is there a way to ignore a single line, like in jshint?
I have tried // tslint ignore:line but that does not seem to work.

### PR DESCRIPTION
**Pull request by patsissons**
This PR supports using ESLint style single line ignore switches by using `tslint:disable-line` and `tslint:disable-next-line`.

This PR resolves #144

### CODE REVIEW

File: src/enableDisableRules.ts [hunk 1]
```diff
@@ -36,25 +37,33 @@ export class EnableDisableRulesWalker extends SkippableTokenAwareRuleWalker {
                 return;
             }
 
+            // keep track of the beginning of the line so that we can use it for same line switches
+            if (scanner.getToken() === ts.SyntaxKind.NewLineTrivia) {
```
Comment (JKillian) [inline src/enableDisableRules.ts]: As opposed to tracking this, you could use:

`node.getLineAndCharacterOfPosition` in conjunction with `scanner.getTokenPos`

File: src/enableDisableRules.ts [hunk 2]
```diff
@@ -63,10 +72,46 @@ export class EnableDisableRulesWalker extends SkippableTokenAwareRuleWalker {
...
+                            position: scanner.getTextPos() + 1,
+                        });
+                    } else {
+                        this.enableDisableRuleMap[ruleToAdd].push({
+                            isEnabled: isEnabled,
+                            position: startingPosition,
+                        });
+
+                        if (isNextLine) {
+                            const nextLineEndPos = scanner.lookAhead(() => {
```
Comment (JKillian) [inline src/enableDisableRules.ts]: I think you should be able to simplify this by passing in the `SourceFile` node and using `node.getPositionOfLineAndCharacter` or maybe `node.getLineStarts`

**Comment by patsissons** [comment]
I'll see what I can come up with tonight

**Comment by JKillian** [comment]
Looking good to me @patsissons! Let's see if we can simplify things a little using the `SourceFile` APIs, but if it seems like a bad idea after trying it to you, we can stick with something more like it is now

**Comment by patsissons** [comment]
I'm all for optimized clean code :+1: what you're suggesting sounds pretty reasonable.

File: src/enableDisableRules.ts [hunk 3]
```diff
@@ -63,10 +72,31 @@ export class EnableDisableRulesWalker extends SkippableTokenAwareRuleWalker {
...
+                    if (isCurrentLine) {
+                        // start at the beginning of the current line
+                        this.enableDisableRuleMap[ruleToAdd].push({
+                            isEnabled: isEnabled,
+                            position: this.getStartOfLinePosition(node, startingPosition),
+                        });
+                        // end at the beginning of the next line
+                        this.enableDisableRuleMap[ruleToAdd].push({
+                            isEnabled: !isEnabled,
+                            position: scanner.getTextPos() + 1,
```
Comment (JKillian) [inline src/enableDisableRules.ts]: Should this be:
`position: this.getStartOfLinePosition(node, startingPosition, 1),` ?

Possible that the current version is fine, I just don't understand what it's doing exactly

**Comment by JKillian** [comment]
👍 Like this new version, much easier to read and understand! Just had one comment above, looks good to me otherwise

File: src/enableDisableRules.ts (hunk 3 above)
Comment (patsissons) [inline src/enableDisableRules.ts]: The third param is line offset, which defaults to 0, the current line. For current line switches we want to start at the beginning of the current line, so line offset isn't necessary.

**Comment by patsissons** [comment]
Thoughts on tests to cover the start and stop interval creation? Or is that better captured in a new PR?

File: src/enableDisableRules.ts (hunk 3 above)
Comment (JKillian) [inline src/enableDisableRules.ts]: Sorry for the lack of clarity, I was referring to if `position: scanner.getTextPos() + 1,` should be changed

**Comment by JKillian** [comment]
As far as tests, I think we should be able to move this over to the newer, more visual testing system. (I.e. the files in `test/rules`. Test case(s) could go in `test/rules/_integration/comments`. That can be a follow-up though

File: src/enableDisableRules.ts (hunk 3 above)
Comment (patsissons) [inline src/enableDisableRules.ts]: Oh sorry, i see what you were trying to say now. Either version should yield the same position. `scanner.getTextPos()` returns the position of the end of the text, the `+1` is just to bump the position to the beginning of the next line. I guess that the current expression may be a little terse, I could add an extra comment for clarity if you like?

Comment (JKillian) [inline src/enableDisableRules.ts]: No worries, the comment above is clear enough! Just wanted to check that this was as intended. PR merged!

**Comment by JKillian** [comment]
@patsissons Do you want to do the test PR? Or should I try to knock it out? Just want to make sure we don't duplicate work

**Comment by patsissons** [comment]
I probably wouldn't get to it for a few days, so go ahead. Plus you'll definitely have a better understanding on how to improve things than I will.

**Pull request merged**